### PR TITLE
Add ddl/schema argument support to get command

### PR DESCRIPTION
**Pull request by hhakimi53**
- Add ddl/schema argument support to get command. By default the sql stored in auto generated ddl file in catalog.jar is stored to the file location provided by user. If not output location is provided, store the file as schema.sql in current working directory.
- Update "get deployment" to output the deployment to deployment.xml in cwd if no output location is specified.

### CODE REVIEW

File: src/frontend/org/voltdb/GetActionArgument.java [hunk 1]
```diff
@@ -19,16 +19,34 @@
...
+    },
+    CLASSES() {
+        @Override
+        public String getDefaultOutput() {
+            return "procedures.jar";
+        }
+    };
 
-    final String m_verb;
+    public String getDefaultOutput() { return ""; }
```
Comment (akhanzode) [inline src/frontend/org/voltdb/GetActionArgument.java]: enum is not the place for this. Move it to VoltDB

File: src/frontend/org/voltdb/GetActionArgument.java [hunk 2]
```diff
@@ -19,16 +19,34 @@
...
 
-    GetActionArgument() {
-        m_verb = name().toLowerCase();
-    }
+    GetActionArgument() { }
 
-    GetActionArgument(String verb) {
-        m_verb = verb;
+    public static String supportedVerbs() {
+        StringBuilder verbNames = new StringBuilder();
```
Comment (akhanzode) [inline src/frontend/org/voltdb/GetActionArgument.java]: Use something like Arrays.asList(values()).toString();

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 3]
```diff
@@ -168,6 +154,23 @@
 import org.voltdb.utils.VoltFile;
 import org.voltdb.utils.VoltSampler;
 
+import org.aeonbits.owner.ConfigFactory;
```
Comment (akhanzode) [inline src/frontend/org/voltdb/RealVoltDB.java]: Dont move imports around.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 4]
```diff
@@ -623,34 +609,87 @@ public void cli(Configuration config) {
...
                 VoltDB.exit(-1);
             }
+            System.out.println("Schema file saved at " + config.m_getOutput.trim());
+        } catch (IOException excp) {
+            System.err.println("Failed to load the catalog jar from " + config.m_pathToCatalog);
+            VoltDB.exit(-1);
+        }
+    }
+
+    void outputProcedure(Configuration config) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/RealVoltDB.java]: Do this in sep ticket.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 5]
```diff
@@ -623,34 +609,87 @@ public void cli(Configuration config) {
...
+        // and Settings depend on
+        ConfigFactory.clearProperty(Settings.CONFIG_DIR);
+        switch (config.m_getOption) {
+            case DEPLOYMENT:
+                outputDeployment(config);
+                break;
+            case SCHEMA:
+                outputSchema(config);
+                break;
+            case CLASSES:
```
Comment (akhanzode) [inline src/frontend/org/voltdb/RealVoltDB.java]: Do it in next ticket.

File: src/frontend/org/voltdb/GetActionArgument.java [hunk 6]
```diff
@@ -19,16 +19,34 @@
...
+            return "deployment.xml";
+        }
+    },
+    SCHEMA() {
+        @Override
+        public String getDefaultOutput() {
+            return "schema.sql";
+        }
+    },
+    CLASSES() {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/GetActionArgument.java]: Do this in next ticket when procedures is implemented.

File: src/frontend/org/voltdb/VoltDB.java [hunk 7]
```diff
@@ -692,6 +722,32 @@ private boolean isInitialized() {
...
+                case DEPLOYMENT: {
+                    File depFH = new VoltFile(configInfoDir, "deployment.xml");
+                    if (!depFH.exists()) {
+                        System.out.println("FATAL: Deployment file \"" + depFH.getAbsolutePath() + "\" not found.");
+                        exit(-1);
+                    }
+                    m_pathToDeployment = depFH.getAbsolutePath();
+                    return;
+                }
+                case SCHEMA:
```
Comment (akhanzode) [inline src/frontend/org/voltdb/VoltDB.java]: Why schema and classes same?

File: src/frontend/org/voltdb/utils/CatalogUtil.java [hunk 8]
```diff
@@ -291,6 +293,25 @@ public static String getSerializedCatalogStringFromJar(InMemoryJarfile jarfile)
...
+     */
+    public static String getAutoGenDDLFromJar(InMemoryJarfile jarfile)
+            throws IOException
+    {
+        // Read the raw build info bytes.
+        byte[] ddlBytes = jarfile.get(VoltCompiler.AUTOGEN_DDL_FILE_NAME);
+        if (ddlBytes == null) {
+            throw new IOException("Auto generated schema DDL not found - please make sure the database is initialized with valid schema.");
+        }
+        String ddl = new String(ddlBytes, Constants.UTF8ENCODING);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: Use StandardCharsets

File: tests/frontend/org/voltdb/regressionsuites/TestInitStartLocalClusterAllOutOfProcess.java [hunk 9]
```diff
@@ -113,23 +122,51 @@ public void testClusterUp() throws Exception
...
+        try {
+            server.cli();
+        } catch (SimulatedExitException ex) {
+            //Good
+        }
 
+        byte[] encoded = Files.readAllBytes(Paths.get(schema.getAbsolutePath()));
+        assertNotNull(encoded);
+        assertTrue(encoded.length > 0);
+        String ddl = new String(encoded, Constants.UTF8ENCODING);
```
Comment (akhanzode) [inline tests/frontend/org/voltdb/regressionsuites/TestInitStartLocalClusterAllOutOfProcess.java]: Use StandardCharsets

File: lib/python/voltcli/voltdb.d/get.py [hunk 10]
```diff
@@ -15,22 +15,40 @@
...
-                   'is voltdbroot under the current working directory.')
 @VOLT.Command(
-    description = 'Get voltdb configuration settings.',
+    description = 'Get voltdb configuration resource.',
     options = (
-        VOLT.StringOption('-o', '--out', 'output', 'File to save configuration obtained.', default=""),
+        VOLT.StringOption('-o', '--output', 'output', 'Specifies file to save the configuration resource to.',
+                          default=None),
+        VOLT.StringOption('-D', '--dir', 'directory_spec', dir_spec_help, default=None),
+        VOLT.BooleanOption('-f', '--force', 'force', 'Stores the output file even if the resouces file from the '
```
Comment (hhakimi53) [inline lib/python/voltcli/voltdb.d/get.py]: Sure.

File: src/frontend/org/voltdb/GetActionArgument.java (hunk 6 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/GetActionArgument.java]: sure

File: src/frontend/org/voltdb/GetActionArgument.java (hunk 2 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/GetActionArgument.java]: ok, thnxs

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 3 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/RealVoltDB.java]: ok

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 4 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/RealVoltDB.java]: ok

File: src/frontend/org/voltdb/utils/CatalogUtil.java (hunk 8 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/utils/CatalogUtil.java]: sure

File: tests/frontend/org/voltdb/regressionsuites/TestInitStartLocalClusterAllOutOfProcess.java (hunk 9 above)
Comment (hhakimi53) [inline tests/frontend/org/voltdb/regressionsuites/TestInitStartLocalClusterAllOutOfProcess.java]: sure

File: src/frontend/org/voltdb/VoltDB.java (hunk 7 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/VoltDB.java]: Both schema and classes require that the catalog.jar to exists, which is needed to fetch auto-generated ddl (schema arg) or to the java stored procedures (class arg). So both of them are evaluated for the same condition.

**Comment by akhanzode** [comment]
Looks good if messages are good and jenkins is happy merge on master.

**Pull request merged**